pkgs/build-support/go/module.nix

```
@@ -10,322 +16,375 @@ lib.extendMkDerivation {
+
+                if [ "$cmd" = "test" ]; then
+                  flags+=(-vet=off)
+                  flags+=($checkFlags)
+                fi
+
+                local OUT
+                if ! OUT="$(go $cmd "''${flags[@]}" $dir 2>&1)"; then
+                  if ! echo "$OUT" | grep -qE '(no( buildable| non-test)?|build constraints exclude all) Go (source )?files'; then
+                    echo "$OUT" >&2
+                    return 1
+                  fi
+                fi
+                if [ -n "$OUT" ]; then
+                  echo "$OUT" >&2
+                fi
+                return 0
+              }
+
+              getGoDirs() {
+                local type;
+                type="$1"
+                if [ -n "$subPackages" ]; then
+                  echo "$subPackages" | sed "s,\(^\| \),\1./,g"
+                else
+                  find . -type f -name \*$type.go -exec dirname {} \; | grep -v "/vendor/" | sort --unique | grep -v "$exclude"
+                fi
+              }
+
+              if [ -z "$enableParallelBuilding" ]; then
+                  export NIX_BUILD_CORES=1
+              fi
+              for pkg in $(getGoDirs ""); do
+                echo "Building subPackage $pkg"
+                buildGoDir install "$pkg"
+              done
+            ''
+          + lib.optionalString (stdenv.hostPlatform != stdenv.buildPlatform) ''
+            # normalize cross-compiled builds w.r.t. native builds
+            (
+              dir=$GOPATH/bin/''${GOOS}_''${GOARCH}
+              if [[ -n "$(shopt -s nullglob; echo $dir/*)" ]]; then
+                mv $dir/* $dir/..
+              fi
+              if [[ -d $dir ]]; then
+                rmdir $dir
+              fi
+            )
+          ''
+          + ''
+            runHook postBuild
+          ''
+        );
+
+      doCheck = args.doCheck or true;
+      checkPhase =
+        args.checkPhase or ''
+          runHook preCheck
+          # We do not set trimpath for tests, in case they reference test assets
+          export GOFLAGS=''${GOFLAGS//-trimpath/}
+
+          for pkg in $(getGoDirs test); do
+            buildGoDir test "$pkg"
+          done
+
+          runHook postCheck
+        '';
+
+      installPhase =
+        args.installPhase or ''
+          runHook preInstall
+
+          mkdir -p $out
+          dir="$GOPATH/bin"
+          [ -e "$dir" ] && cp -r $dir $out
+
+          runHook postInstall
+        '';
+
+      strictDeps = true;
+
+      disallowedReferences = lib.optional (!allowGoReference) go;
+
+      passthru = {
+        inherit go;
+        # Canonicallize `overrideModAttrs` as an attribute overlay.
+        # `passthru.overrideModAttrs` will be overridden
+        # when users want to override `goModules`.
+        overrideModAttrs = lib.toExtension overrideModAttrs;
+      } // passthru;
+
+      meta = {
+        # Add default meta information.
+        platforms = go.meta.platforms or lib.platforms.all;
```

For a future PR: I suspect that everything after this `or` is dead code.

```suggestion
        inherit (go.meta) platforms;
```